lighthouse-core/fraggle-rock/user-flow.js

```diff
@@ -176,6 +166,69 @@ class UserFlow {
     const flowResult = await this.createFlowResult();
     return generateFlowReportHtml(flowResult);
   }
+
+  /**
+   * @return {LH.UserFlow.FlowArtifacts}
+   */
+  createArtifactsJson() {
+    /** @type {LH.UserFlow.GatherStep[]} */
+    const stepArtifacts = this._gatherSteps.map(step => ({
+      name: step.name,
+      artifacts: step.artifacts,
+      config: step.config,
+      configContext: step.configContext,
+    }));
+
+    return {
+      gatherSteps: stepArtifacts,
+      name: this.name,
+    };
+  }
+}
+
+/**
+ * @param {Array<LH.UserFlow.GatherStep | LH.UserFlow.ActiveGatherStep>} gatherSteps
+ * @param {string|undefined} name
+ * @param {LH.Config.Json|undefined} flowConfig Must be the same as the config provided when the flow started.
+ */
+async function auditGatherSteps(gatherSteps, name, flowConfig) {
+  if (!gatherSteps.length) {
+    throw new Error('Need at least one step before getting the result');
+  }
+
+  /** @type {LH.FlowResult['steps']} */
+  const steps = [];
+  for (const gatherStep of gatherSteps) {
+    const {artifacts, name, configContext} = gatherStep;
+
+    let runnerOptions;
+    if ('runnerOptions' in gatherStep) {
+      // If the gather step is active, we can access the runner options directly.
+      runnerOptions = gatherStep.runnerOptions;
+    } else {
+      // If the gather step is not active, we must recreate the runner options.
+      // Step specific configs take precedence over a config for the entire flow.
+      const configJson = gatherStep.config || flowConfig;
+      const {gatherMode} = artifacts.GatherContext;
+      const {config} = initializeConfig(configJson, {...configContext, gatherMode});
+      runnerOptions = {
```

Comment:
seems we could just maintained a second property like `_gatherStepRunnerOptions` or something, and keep runner options data there. could be an array or a weakmap (if array, this loop changes to `for (let i ...`, if weakmap can keep for loop as is). I think conceptually the weakmap makes sense here.

Reply:
I decided to go with the weak map approach here. Keeps `GatherStep` type clear of any extra stuff